server/server/src/billing/billing_service.rs

```diff
@@ -73,6 +79,67 @@ fn release_subscription_profile<T: Debug>(
     })?)
 }
 
+pub async fn upgrade_account_app_store(
+    context: RequestContext<'_, UpgradeAccountAppStoreRequest>,
+) -> Result<UpgradeAccountAppStoreResponse, ServerError<UpgradeAccountAppStoreError>> {
+    let (request, server_state) = (&context.request, context.server_state);
+
+    let mut account = lock_subscription_profile(server_state, &context.public_key)?;
+
+    if account.billing_info.data_cap() == PREMIUM_TIER_USAGE_SIZE {
+        return Err(ClientError(UpgradeAccountAppStoreError::AlreadyPremium));
+    }
+
+    debug!("Upgrading the account of a user through app store billing");
+
+    if let Some(owner) = server_state
+        .index_db
+        .app_store_ids
+        .get(&request.app_account_token)?
+    {
+        if let Some(other_account) = server_state.index_db.accounts.get(&owner)? {
+            if other_account.billing_info.billing_platform.is_some() {
+                return Err(ClientError(UpgradeAccountAppStoreError::AppStoreAccountAlreadyLinked));
```

Comment:
wouldn't we return this error if they had `Strip`? Prob need to rename the error

Reply:
I ended up `if let`ing an app store subscription, so there is no need to rename. Also, it is what should be done since we only care if it is an app store subscription.

Reply:
What if they have another platform like stripe?

Reply:
It doesn't matter if they have another platform so long as the platform is not apple. But on second thought, it is possible that the platform is apple but not the same apple account. I will add a check for the origin transaction id.

If the account happened to be stripe it would not matter because all this check is looking for is if an apple account is not being used by someone else. An apple account can only pay for one subscription.